src/azure-cli-core/azure/cli/core/cloud.py

```diff
@@ -454,18 +467,45 @@ def __str__(self):
         postgresql_server_endpoint='.postgres.database.cloudapi.de',
         mariadb_server_endpoint='.mariadb.database.cloudapi.de'))
 
-KNOWN_CLOUDS = [AZURE_PUBLIC_CLOUD, AZURE_CHINA_CLOUD, AZURE_US_GOV_CLOUD, AZURE_GERMAN_CLOUD]
 
-if 'ARM_CLOUD_METADATA_URL' in os.environ:
-    try:
-        arm_cloud_dict = json.loads(urlretrieve(os.getenv('ARM_CLOUD_METADATA_URL')))
-        cli_cloud_dict = _convert_arm_to_cli(arm_cloud_dict)
-        if 'AzureCloud' in cli_cloud_dict:
-            cli_cloud_dict['AzureCloud'].endpoints.active_directory = 'https://login.microsoftonline.com'  # change once active_directory is fixed in ARM for the public cloud
-        KNOWN_CLOUDS = list(cli_cloud_dict.values())
-    except Exception as ex:  # pylint: disable=broad-except
-        logger.warning('Failed to load cloud metadata from the url specified by ARM_CLOUD_METADATA_URL')
-        raise ex
+def get_known_clouds():
+    if 'ARM_CLOUD_METADATA_URL' in os.environ:
+        from azure.cli.core._session import CLOUD_ENDPOINTS
+        endpoints_file = os.path.join(GLOBAL_CONFIG_DIR, 'cloudEndpoints.json')
+        CLOUD_ENDPOINTS.load(endpoints_file)
+        clouds = []
+        if CLOUD_ENDPOINTS['clouds']:
```

Comment:
if user set ARM_METADATA_URL and config file contains 'clouds' setting, that means clouds will read from meta url firstly, then overwrite it with local config `clouds` section. Is this right? shouldn't it from metadata url directly? then update into file?

Reply:
The logic is to first load `KNOWN_CLOUDS`, then load the config 'clouds' settings (active cloud, extra registered clouds, modified endpoints). This mechanism will not change. With ARM_METADATA_URL, KNOWN_CLOUDS is obtained through the URL or the `cloudEndpoints.json`. It does not affect the config 'clouds' settings part.

Reply:
Another thing changed is that previously when there is no active cloud set in config file, it will use `AzureCloud`, but now we check if `AzureCloud` is in `KNOWN_CLOUDS`. If not, we set the first one in `KNOWN_CLOUDS` as active in config file.